### ISSUE TITLE
sessionImpl.mux can be nil

### ISSUE THREAD

**Issue by ktis** [icza/session#5]
Second request(reload) made runtime error in my project.

> 2017/03/04 03:16:12 http: panic serving 127.0.0.1:60332: runtime error: invalid memory address or nil pointer dereference
goroutine 42 [running]:
net/http.(*conn).serve.func1(0xc820244180)
	/usr/local/go_appengine/goroot/src/net/http/server.go:1389 +0xc1
panic(0x93e320, 0xc82000a070)
	/usr/local/go_appengine/goroot/src/runtime/panic.go:443 +0x4e9
sync.(*RWMutex).Lock(0x0)
	/usr/local/go_appengine/goroot/src/sync/rwmutex.go:83 +0xc6
github.com/icza/session.(*sessionImpl).Access(0xc82004f680)
	/usr/local/go_appengine/src/github.com/icza/session/session.go:227 +0x29
github.com/icza/session.(*memcacheStore).Get(0xc82024a1e0, 0xc820167427, 0x18, 0x0, 0x0)

I try to fix mux-type and related code, then problem was solved.
session.go line80
Before
`mux       *sync.RWMutex`
After
`mux       sync.RWMutex`

If it is bug, I hope fix.

**Comment by icza** [comment]
I will check it later, meanwhile can you send me the code how you use this `session` package?

**Comment by ktis** [comment]
I use session package as below.
Firstly I request with login parameter, then func Login called and new session added.
Secondly I request with the sid, then func Toppage called and getSessionID failed.

```golang
import (
        oldappengine "appengine" 
       "github.com/icza/session"
}

func makeSession(w http.ResponseWriter, r *http.Request) {
        c2 := oldappengine.NewContext(r)
        sessmgr := session.NewCookieManagerOptions(session.NewMemcacheStore(c2), &session.CookieMngrOptions{AllowHTTP: true})
        defer sessmgr.Close() // Note the Close(): it will ensure changes made to the session are auto-saved in Memcache.
        uid := r.FormValue("uid")       
        // New session with initial constant and variable attributes:
        sess := session.NewSessionOptions(&session.SessOptions{
                CAttrs: map[string]interface{}{"uid": uid},
        })
        sess.SetAttr("Count", 0)        
        sessmgr.Add(sess, w)
}

func Login(ctx context.Context, w http.ResponseWriter, r *http.Request) { // Login handler
        status := r.FormValue("status")     
        if status == "login" {        
                makeSession(w, r)               
        } else if status == "logout" {  
                removeSession(w, r)
        }
}

func getSessionID(w http.ResponseWriter, r *http.Request) (sessionID string, uid string) {
        c2 := oldappengine.NewContext(r)
        sessmgr := session.NewCookieManagerOptions(session.NewMemcacheStore(c2), &session.CookieMngrOptions{AllowHTTP: true})
        defer sessmgr.Close() // Note the Close(): it will ensure changes made to the session are auto-saved in Memcache.
        sess := sessmgr.Get(r)//Second request made error.
        if sess != nil {
                // Already logged in
                sess.SetAttr("Count", sess.Attr("Count").(int)+1)
        } else {
                return "", ""
        }   
        return sess.ID(), sess.CAttr("uid").(string)
}

func Toppage(c context.Context, w http.ResponseWriter, r *http.Request) {//Toppage handldr. When http request received, this function is called.
        sessionID, uid := getSessionID(w, r)
}
```

**Comment by icza** [comment]
There is some conceptual misunderstanding how you use the session manager.
`SessionManager.Close()` should only be closed when the request ends.

In the example `defer sessmgr.Close()` is used which ensures that when the request ends, any changes you made to the session (e.g. added, changed or removed attributes) are "persisted" to the memcache.

You should not call it at "random" times. When a request comes in (or when you first need the session), create the session manager, but do not close it if there's a chance you need to access it again during serving the request.

**Comment by ktis** [comment]
Thanks.
However, problem is not solved after fixing about `defer sessmgr.Close()` 

Then I try to run gae_session_demo, and same problem occured when click 'reflesh' button.
The log is below.
I use appengine latest version  "1.9.48".   

> 2017/03/11 11:30:24 Session demo is about to start. Visit: localhost:8080/demo
2017/03/11 11:30:53 INFO: Session added: 38e7l0UWyTPeOZUkVXyPBGyO
2017/03/11 11:30:57 http: panic serving 127.0.0.1:43374: runtime error: invalid memory address or nil pointer dereference
goroutine 15 [running]:
net/http.(*conn).serve.func1(0xc820074c00)
	/usr/local/go_appengine/goroot/src/net/http/server.go:1389 +0xc1
panic(0x8c44c0, 0xc82000a070)
	/usr/local/go_appengine/goroot/src/runtime/panic.go:443 +0x4e9
sync.(*RWMutex).Lock(0x0)
	/usr/local/go_appengine/goroot/src/sync/rwmutex.go:83 +0xc6
github.com/icza/session.(*sessionImpl).Access(0xc820183020)
	/usr/local/go_appengine/src/github.com/icza/session/session.go:227 +0x29
github.com/icza/session.(*memcacheStore).Get(0xc820182fc0, 0xc820185407, 0x18, 0x0, 0x0)
	/usr/local/go_appengine/src/github.com/icza/session/gae_memcache_store.go:215 +0x81d
github.com/icza/session.(*CookieManager).Get(0xc820187740, 0xc820108a80, 0x0, 0x0)
	/usr/local/go_appengine/src/github.com/icza/session/cookie_manager.go:82 +0xca
main41877.myHandler(0x7f358de50320, 0xc8202044e0, 0xc820108a80)
	gae_session_demo.go:40 +0x31d
net/http.HandlerFunc.ServeHTTP(0xa75788, 0x7f358de50320, 0xc8202044e0, 0xc820108a80)
	/usr/local/go_appengine/goroot/src/net/http/server.go:1618 +0x3a
net/http.(*ServeMux).ServeHTTP(0xc82000eab0, 0x7f358de50320, 0xc8202044e0, 0xc820108a80)
	/usr/local/go_appengine/goroot/src/net/http/server.go:1910 +0x17d
appengine_internal.handleFilteredHTTP(0x7f358de50320, 0xc8202044e0, 0xc820108a80)
	/usr/local/go_appengine/goroot/src/appengine_internal/api_dev.go:105 +0x44c
net/http.HandlerFunc.ServeHTTP(0xa74d08, 0x7f358de50320, 0xc8202044e0, 0xc820108a80)
	/usr/local/go_appengine/goroot/src/net/http/server.go:1618 +0x3a
net/http.serverHandler.ServeHTTP(0xc82019c100, 0x7f358de50320, 0xc8202044e0, 0xc820108a80)
	/usr/local/go_appengine/goroot/src/net/http/server.go:2081 +0x19e
net/http.(*conn).serve(0xc820074c00)
	/usr/local/go_appengine/goroot/src/net/http/server.go:1472 +0xf2e
created by net/http.(*Server).Serve
	/usr/local/go_appengine/goroot/src/net/http/server.go:2137 +0x44e

**Comment by icza** [comment]
There was indeed a bug in the implementation, which is fixed by commit 2eda5d157a03149bf944ca6ce8e27592da98f364. Closing this issue.
